### PR TITLE
Enhance search module and UI

### LINKED ISSUE #13239: Enhance Search module and make the UI configurable
### Is your feature request related to a problem? Please describe.
Currently, we have two different `SearchController`. each one handles it's own search provider Licence and ElasticSearch. We should be able to combine these two into one and place it on the `Search` module since that module's sole purpose is to export a front end UI for the search. Also, we should add a setting to change the placeholder for the search bar. Additionally, we should have `Search` widget that would allow the user to add search widget.

### PR DESCRIPTION
**Pull request by MikeAlhayek**
This PR Fix #13239 
1. Remove the `SearchController` from both Elasticsearch in Lucene module. We now have one common `SearchController` in the `Search` module.
2. The new controller now returns a Shape instead of a view-model which allows the user to override the default views.
3. Adds settings to allow the user to defined title for the search page and also a placeholder text for the search bar.
4. Add Search widget that a user can add to their site. The default widget will direct the user to the `/search` route.
5. Allow the user to optionally pass index name in the route value. For example, `/search/index-name` the index-name is an optional value and when it is not provided, the default index is used.

### CODE REVIEW

**Comment by MikeAlhayek** [comment]
@Skrypt this one is for you to review when you have time.

**Comment by Skrypt** [comment]
Looking good. Just see small things here and there. Will do the review this week.

File: src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+        S = stringLocalizer;
+        _serviceProvider = serviceProvider;
+        _shapeFactory = shapeFactory;
+        New = shapeFactory;
+        _logger = logger;
+    }
+
+    public async Task<IActionResult> Search(string index, SearchIndexViewModel viewModel, PagerSlimParameters pagerParameters)
+    {
+        if (!String.IsNullOrEmpty(index))
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs:59]: Questioning the reason for this parameter

File: src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+            _logger.LogInformation("No search provider feature is enabled.");
+
+            return BadRequest("Missing search provider service.");
+        }
+
+        var site = await _siteService.GetSiteSettingsAsync();
+
+        var searchSettings = site.As<SearchSettings>();
+
+        var searchService = GetSearchService(searchServices, searchSettings?.SearchProviderAreaName);
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs:77]: Why AreaName (legacy name)

File: src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        var indexName = !String.IsNullOrWhiteSpace(viewModel.Index) ? viewModel.Index.Trim() : await searchService.DefaultIndexAsync();
+
+        if (indexName == null || !await searchService.ExistsAsync(indexName))
+        {
+            _logger.LogInformation("Couldn't execute search. The search index doesn't exist.");
+
+            return BadRequest("Search provider is not configured.");
+        }
+
+        var permission = await searchService.GetPermissionAsync(indexName);
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs:88]: Search permission, common to all providers -> abstraction

File: src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+
+        var searchService = GetSearchService(searchServices, searchSettings?.SearchProviderAreaName);
+
+        var indexName = !String.IsNullOrWhiteSpace(viewModel.Index) ? viewModel.Index.Trim() : await searchService.DefaultIndexAsync();
+
+        if (indexName == null || !await searchService.ExistsAsync(indexName))
+        {
+            _logger.LogInformation("Couldn't execute search. The search index doesn't exist.");
+
+            return BadRequest("Search provider is not configured.");
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs:85]: Log configuration issues as warning, return no results.

File: src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+        }
+
+        var permission = await searchService.GetPermissionAsync(indexName);
+
+        if (permission != null && !await _authorizationService.AuthorizeAsync(User, permission))
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs:90]: Use common permission from Search module, each implementation of the search can "change" which one is checked with a custom handler, even from dynamic permissions.

Comment (sebastienros) [reply src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs:90]: Pass a custom SearchPermissionParameters(Index, IUser) such that the engine can return the specific result.

The engines could check if they are the search engine for the service.

File: src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs (hunk 1 above)
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs]: Done!

File: src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs (hunk 3 above)
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs]: Done!

File: src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs (hunk 5 above)
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs]: Done!

File: src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs (hunk 4 above)
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs]: Done!

File: src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs (hunk 2 above)
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs]: this may be outside the scope of this PR.

**Comment by MikeAlhayek** [comment]
> Looking good. Just see small things here and there. Will do the review this week.

@Skrypt made some changes based on @sebastienros feedback. Let me know your thoughts.

**Comment by MikeAlhayek** [comment]
@sebastienros I made the requested changes. I'll await until tomorrow just in case you have additional feedback on this.

File: src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+                .ToList(),
+            };
+            model.Pager = (await New.PagerSlim(pager)).UrlParams(new Dictionary<string, string>()
+            {
+                { "Terms", viewModel.Terms },
+                { "Index", viewModel.Index },
+            });
+        });
+
+        return View(shape);
```
Comment (jtkech) [inline src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs:171]: @MikeAlhayek @sebastienros Just for info we already can override a pure mvc view in a theme without having to create a shape with the `_shapeFactory`, from memory under `Views/ModuleArea/ControllerName/Action.cshtml`.

Or to use a simpler path as for shape, maybe just make `SearchIndexViewModel` inheriting from `ShapeViewModel`, as we already did for `SearchFormViewModel` and `SearchResultsViewModel`.

Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Search/Controllers/SearchController.cs:171]: just confirmed that `Views/ModuleArea/ControllerName/Action.cshtml` is the right path too.

**Pull request merged**